components/daffy/daffy.app.mjs

```diff
@@ -0,0 +1,138 @@
+import { axios } from "@pipedream/platform";
+import { defineApp } from "@pipedream/types";
+
+export default defineApp({
+  type: "app",
+  app: "daffy",
+  propDefinitions: {
+    causeId: {
+      type: "string",
+      label: "Cause",
+      description: "The ID of the main cause.",
+      async options() {
+        const data = await this.listCauses();
+
+        return data.map(({
+          id: value, name: label,
+        }) => ({
+          label,
+          value,
+        }));
+      },
+    },
+    ein: {
+      type: "string",
+      label: "EIN",
+      description: "The Employer Identification Number of the non-profit.",
+      async options({
+        page, causeId,
+      }) {
+        const { items: data } = await this.listNonProfits({
+          params: {
+            cause_id: causeId,
+            page: page + 1,
+          },
+        });
+
+        return data.map(({
+          ein: value, name: label,
+        }) => ({
+          label,
+          value,
+        }));
+      },
+    },
+  },
+  methods: {
+    _apiUrl() {
+      return "https://public.daffy.org/v1";
+    },
+    _getHeaders() {
+      return {
+        "X-Api-Key": this.$auth.api_key,
+      };
+    },
+    async _makeRequest({
+      $ = this, path, ...opts
+    }) {
+      const config = {
+        url: `${this._apiUrl()}/${path}`,
+        headers: this._getHeaders(),
+        ...opts,
+      };
+
+      return axios($, config);
+    },
+    async createDonation(args = {}) {
+      const userId = await this.getMe();
+      return this._makeRequest({
+        method: "POST",
+        path: `users/${userId}/donations`,
+        ...args,
+      });
+    },
+    listNonProfits(args = {}) {
+      return this._makeRequest({
+        path: "non_profits",
+        ...args,
+      });
+    },
+    async listCauses(args = {}) {
+      const userId = await this.getMe();
+      return this._makeRequest({
+        path: `users/${userId}/causes`,
+        ...args,
+      });
+    },
+    async listDonations(args = {}) {
+      const userId = await this.getMe();
+      return this._makeRequest({
+        path: `users/${userId}/donations`,
+        ...args,
+      });
+    },
+    async getMe(args = {}) {
+      const { id: userId } = await this._makeRequest({
+        path: "users/me",
+        ...args,
+      });
+
+      return userId;
+    },
+    async *paginate({
+      fn, params = {}, maxResults = null,
+    }) {
+      let firstPage = false;
+      let count = 0;
+      let page = 0;
+      let firstRequest = 1;
+
+      do {
+        params.page = page--;
```

Comment:
`page` will be -1 the first time through. Is that correct?

Reply:
This API does not allow us to order the request, so I created a pagination that the first time is used only to get the lastPage then the next loops will take the last page and decrement it. so in fact the page must start = 1, page-- will assign the value 1 to params.page and only then change the value to 0, but then it no longer matters because in the sequence the page receives the value of lastPage